text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+enabled. The other must _not_ be enabled, meaning they are mutually exclusive.
+
+#### `hasDestroyable`
+
+Determines if the helper has a destroyable to include in the destructor
+hierarchy. If enabled, the `getDestroyable` hook will be called, and its result
+will be associated with the destroyable parent block.
+
+#### `hasValue`
+
+Determines if the helper has a value which can be used externally. The helper's
+`getValue` hook will be run whenever the value of the helper is accessed if this
+capability is enabled.
+
+#### `isScheduledEffect`
+
+Determines if the helper is scheduled. If enabled, the helper's `runEffect` hook
+will run after render, and will not allow any type of state mutation when
+running.
+
+### Scheduled Helpers Timing
+
+Scheduled helpers run their effects after render, and after modifiers have been
+applied for a given render, but before paint. The exact timing may shift around,
+and may or may not correspond to a single rendering pass in cases where there
+are multiple rendering passes in a single paint.
+
+In the future different timings may be added as options for scheduling. For
+instance, a timing to call the effect using `requestIdleCallback`, when the
+browser has finished rendering and handling higher priority work, could be
+added. However, this is out of scope for this RFC.
+
+## How we teach this
+
+Helper managers are a low-level construct that is generally only meant to be
+used by experts and addon authors. As such, it will only be taught through API
+documentation. In addition, for precision and clarity, the API docs will include
+snippets of TypeScript interfaces where appropriate.
+
+### API Docs
+
+#### `setHelperManager`
+
+Sets the helper manager for an object or function.
+
+```js
+setHelperManager((owner) => new ClassHelperManager(owner), Helper.prototype)
+```
+
+When a value is used as a helper in a template, the helper manager is looked up
+by on the object by walking up its prototype chain and finding the first helper
+manager. This manager then receives the value and can create and manage an
+instance of a helper from it. This provides a layer of indirection that allows
+users to design high-level helper APIs, without Ember needing to worry about the
+details. High-level APIs can be experimented with and iterated on while the
+core of Ember helpers remains stable, and new APIs can be introduced gradually
+over time to existing code bases.
+
+`setHelperManager` receives two arguments:
+
+1. A factory function, which receives the `owner` and returns an instance of a
+   helper manager.
+2. The object to associate the factory function with.
+
+The first time the object is looked up, the factory function will be called to
+create the helper manager. It will be cached, and in subsequent lookups it the
+cached helper manager will be used instead.
+
+Only one helper manager exists per helper factory, so many helpers will end up
```

Comment:
> per helper factory

I don't think the term "helper factory" is actually defined here (this is the only reference in the whole document). Would helper definition work here?

Reply:
hmm, I think it was supposed to be "one helper manager per helper _manager_ factory". The semantics are that we weakmap the function that produces the manager to the instance of the manager, so you could use the same factory function with many different definitions, and they would all use the same instance. That's the thing we need to capture here.

Reply:
Ya, my concern wasn't about that coupling, but it was with the terminology used to talk about which thing the factory function is associated with. There is no conceptual thing introduced in this RFC as a "helper factory". There is a `HelperDefinition` (which is what this argument is actually called throughout the rest of this document), but no "helper factory". 

Also, conceptually calling this a "helper factory" is misleading. There is no reason for us to infer that this value is a "factory" at all (factory usually inferring that you can do `new Foo()` or `Foo.create()` in Ember parlance). For example, it is perfectly valid for this value to be `{}`.

Reply:
right, but again, I'm not talking about the definition, I'm talking about the function that creates the helper manager itself. That is a different thing entirely, which we need a new term for, it is _not_ the `HelperDefinition`

Reply:
Hmm, why would we do a different thing here than for component managers / modifier managers? We do not cache the factory instance on the manager factory function, we call the manager factory function for each component base class we find.

See this demo repo if you'd like to fact check me 😜:

https://github.com/rwjblue/manager-invocation-demo

Reply:
Ah, this is my mistake then. This is how managers work in Glimmer.js currently, I must've forgotten where we put that optimization. It looks like it first came from `ember-functional-modifiers`, which makes sense because that addon needed to assign the same modifier manager to many functions and we wanted to avoid creating an instance per function: https://github.com/ember-modifier/ember-modifier/blob/master/addon/-private/functional/modifier.js

Personally I think this optimization is common enough that we should consider making it the default, but we can make it something users have to implement if they need it. I'm not sure if it would be a breaking change to update the existing managers - in theory, they shouldn't depend on instance state per-definition (they may depend on state overall, or per owner), that would somewhat defeat the purpose.

Reply:
> Personally I think this optimization is common enough that we should consider making it the default

Ya, I'm sure. I think that might be fine, but adding caching for something that is super light weight (`new Something()` where that constructor does _basically_ nothing) is actually going to make this slower a lot of times.

Even if we do this, the prose here is wrong. It is not one per manager factory function, it is one per manager factory function per owner. I think that is conceptually fine though, just needs tweaking to be clearer I guess.

Reply:
If we do that, we should also consider doing the same for the other manager types. I'd hate to introduce another small difference while our overall goal is to _reduce_ the differences. I think we can just do the implementation change for the others though, I don't particularly think it is a breaking change or anything.